src/faidx/mod.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 Manuel Landesfeind, Evotec International GmbH
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//!
+//! Module for working with faidx-indexed FASTA files.
+//!
+
+use std::ffi;
+use std::path::Path;
+use url::Url;
+
+use crate::htslib;
+
+pub mod errors;
+pub use errors::{Error, Result};
+
+fn path_as_bytes<'a, P: 'a + AsRef<Path>>(path: P, must_exist: bool) -> Result<Vec<u8>> {
+    if path.as_ref().exists() || !must_exist {
+        Ok(path
+            .as_ref()
+            .to_str()
+            .ok_or(Error::NonUnicodePath)?
+            .as_bytes()
+            .to_owned())
+    } else {
+        Err(Error::FileNotFound {
+            path: path.as_ref().to_owned(),
+        })
+    }
+}
+/// A Fasta reader.
+#[derive(Debug)]
+pub struct Reader {
+    inner: *mut htslib::faidx_t,
+}
+
+impl Reader {
+    /// Create a new Reader from path.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path to open.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
+        Self::new(&path_as_bytes(path, true)?)
+    }
+
+    /// Create a new Reader from URL.
+    pub fn from_url(url: &Url) -> Result<Self, Error> {
+        Self::new(url.as_str().as_bytes())
+    }
+
+    /// Create a new Reader.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path to open
+    fn new(path: &[u8]) -> Result<Self, Error> {
+        let cpath = ffi::CString::new(path).unwrap();
+        let inner = unsafe { htslib::fai_load(cpath.as_ptr()) };
+        Ok(Self { inner })
+    }
```

Comment:
What is the distinction between this `new()` function to create a `Reader` and the `from_path()` function above? From what I can see, only the `from_path()` above is being tested below. So maybe this is an earlier attempt? Or does it provide another benefit and should also be tested?

Reply:
I agree. But I followed the general pattern of the lib. For example, there is also `bam/errors.rs` etc. If all maintainer agree, then we can start combining this.